File: RottingApples/RottingApples.cpp
// RottingApples.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <vector>
#include <queue>

int GetRotDays(std::vector<std::vector<int>>& grid, std::queue<std::pair<size_t, size_t>>& rots)
{
    if (rots.size() == 0)
        return -1;

    std::queue<std::pair<size_t, size_t>> next;

    while (!rots.empty())
    {
        auto [r, c] = rots.front();
        if (r == grid.size() - 1 && c == grid[0].size() - 1)
            return 0;

        rots.pop();
        if (r != 0 && grid[r - 1][c] == 1)
        {
            grid[r - 1][c] = 2;
            next.push({ r - 1, c });
        }
        if (c != 0 && grid[r][c - 1] == 1)
        {
            grid[r][c - 1] = 2;
            next.push({ r, c - 1 });
        }
        if (r != grid.size() - 1 && grid[r + 1][c] == 1)
        {
            grid[r + 1][c] = 2;
            next.push({ r + 1, c });
        }
        if (c != grid[0].size() - 1 && grid[r][c + 1] == 1)
        {
            grid[r][c + 1] = 2;
            next.push({ r, c + 1 });
        }
    }

    auto days = GetRotDays(grid, next);
    if (days == -1)
        return days;
    return 1 + days;
}

int GetDaysToRot(std::vector<std::vector<int>>& grid)
{
    std::vector<std::vector<int>> gridCopy(grid);  // copy the grid so that it won't change the original grid
    std::queue<std::pair<size_t, size_t>> rots;
    rots.push({ 0, 0 });

    return GetRotDays(gridCopy,rots);
}

int main()
{
    while (true)
    {
        int rows = 0, columns = 0;
        std::cout << "Rows and Columns of the grid: ";
        std::cin >> rows >> columns;
        if (rows == 0 && columns == 0)
            break;

        std::vector<std::vector<int>> grid(rows, std::vector<int>(columns, 0));
        std::cout << "Enter the grid: ";
        for (int i = 0; i < rows; ++i)
        {
            for (int j = 0; j < columns; ++j)
                std::cin >> grid[i][j];
        }

        auto days = GetDaysToRot(grid);
        if (days != -1)
            std::cout << "It takes " << GetDaysToRot(grid) << " days to rot all the apples" << std::endl;
        else
            std::cout << "It is impossible to rot all the apples!" << std::endl;

        std::cout << std::endl;
    }
}

File: Go/RottingApples.go
package main

import "fmt"

type Cell struct {
	i, j int
}

func DaysToRot(grid [][]int, rots []Cell) int {
	if len(rots) == 0 {
		return -1
	}
	var newRots []Cell
	for i := 0; i < len(rots); i++ {
		c := rots[i]
		if c.i == len(grid)-1 && c.j == len(grid[0])-1 {
			return 0
		}

		if c.i != 0 && grid[c.i-1][c.j] == 1 {
			grid[c.i-1][c.j] = 2
			newRots = append(newRots, Cell{c.i - 1, c.j})
		}

		if c.j != 0 && grid[c.i][c.j-1] == 1 {
			grid[c.i][c.j-1] = 2
			newRots = append(newRots, Cell{c.i, c.j - 1})
		}

		if c.i != len(grid)-1 && grid[c.i+1][c.j] == 1 {
			grid[c.i+1][c.j] = 2
			newRots = append(newRots, Cell{c.i + 1, c.j})
		}

		if c.j != len(grid[0])-1 && grid[c.i][c.j+1] == 1 {
			grid[c.i][c.j+1] = 2
			newRots = append(newRots, Cell{c.i, c.j + 1})
		}
	}

	d := DaysToRot(grid, newRots)
	if d == -1 {
		return -1
	}
	return d + 1
}

func GetDaysToRot(grid [][]int) int {
	rots := make([]Cell, 1)
	rots[0] = Cell{0, 0}

	return DaysToRot(grid, rots)
}

func main() {
	for {
		fmt.Print("Please enter the size of the grid (rows, cols): ")
		var rows, cols int
		fmt.Scan(&rows, &cols)
		if rows == 0 && cols == 0 {
			break
		}

		grid := make([][]int, rows)
		for r := 0; r < rows; r++ {
			fmt.Print("Please enter the ")
			fmt.Print(r)
			fmt.Print("th row of the grid: ")
			grid[r] = make([]int, rows)
			for c := 0; c < cols; c++ {
				fmt.Scan(&grid[r][c])
			}
		}

		days := GetDaysToRot(grid)
		if days == -1 {
			fmt.Println("It's impossible to rot all the apples!")
		} else {
			fmt.Print("It takes ")
			fmt.Print(days)
			fmt.Println(" days to rot all the apples.")
		}
		fmt.Println("")
	}
}
